src/backend/web/handlers/webhooks.py

```diff
@@ -0,0 +1,140 @@
+import uuid
+
+from flask import Blueprint, redirect, request, url_for
+from google.appengine.ext import ndb
+from pyre_extensions import none_throws
+
+from backend.common.auth import current_user
+from backend.common.consts.client_type import ClientType
+from backend.common.helpers.tbans_helper import TBANSHelper
+from backend.common.models.account import Account
+from backend.common.models.mobile_client import MobileClient
+from backend.web.decorators import require_login
+from backend.web.profiled_render import render_template
+
+blueprint = Blueprint("webhooks", __name__, url_prefix="/webhooks")
+
+
+@blueprint.route("/add", methods=["GET"])
+@require_login
+def webhook_add():
+    template_values = {
+        "error": request.args.get("error"),
+    }
+
+    return render_template("webhook_add.html", template_values)
+
+
+@blueprint.route("/add", methods=["POST"])
+@require_login
+def webhook_add_post():
+    url = request.form.get("url")
+    name = request.form.get("name")
+
+    if not url or not name:
+        return url_for(".webhook_add", error=1)
+
+    # Always generate secret server-side; previously allowed clients to set the secret
+    secret = uuid.uuid4().hex
+
+    user = none_throws(current_user())
+    query = MobileClient.query(
+        MobileClient.messaging_id == url,
+        ancestor=user.account_key,
+    )
+    if query.count() == 0:
+        # Webhook doesn't exist, add it
+        verification_key = TBANSHelper.verify_webhook(url, secret)
+
+        client = MobileClient(
+            parent=user.account_key,
+            user_id=str(user.uid),
+            messaging_id=url,
+            display_name=name,
+            secret=secret,
+            client_type=ClientType.WEBHOOK,
+            verified=False,
+            verification_code=verification_key,
+        )
+        client.put()
+    else:
+        # Webhook already exists. Update the secret
+        current = query.fetch()[0]
+        current.secret = secret
+        current.put()
+
+    return redirect(url_for("account.overview"))
+
+
+@blueprint.route("/delete", methods=["POST"])
+@require_login
+def webhook_delete():
+    user = none_throws(current_user())
+
+    client_id = request.form.get("client_id")
+    if not client_id:
+        return redirect(url_for("account.overview"))
+
+    to_delete = ndb.Key(Account, user.account_key.id(), MobileClient, int(client_id))
+    to_delete.delete()
+
+    return redirect(url_for("account.overview"))
+
+
+@blueprint.route("/verify/<int:client_id>", methods=["GET"])
+@require_login
+def webhook_verify(client_id):
+    template_values = {"error": request.args.get("error"), "client_id": client_id}
```

Comment:
Same note here around using the session as opposed to sending error values through args